vllm/distributed/kv_transfer/kv_connector/v1/cpu_connector.py

```diff
@@ -0,0 +1,936 @@
+    Args:
+        src_layer (torch.Tensor): The source layer on device, shape is 
+            (2, num_vllm_blocks, page_size, ...remaining dims...) 
+        dst_buffer (torch.Tensor): The destination buffer on host, shape is
+            (2, len(block_ids), page_size, ...remaining dims...)
+        block_ids (list[int]): The list of vllm block ids to copy from.
+    """
+    block_mapping = torch.stack([
+        torch.tensor(block_ids, dtype=torch.long),
+        torch.arange(len(block_ids), dtype=torch.long)
+    ],
+                                dim=1)
+    ops.swap_blocks(src_layer[0], dst_buffer[0], block_mapping)
+    ops.swap_blocks(src_layer[1], dst_buffer[1], block_mapping)
+
+
+def h2d_copy_leading_tokens(src_buffer: torch.Tensor, dst_layer: torch.Tensor,
+                            src_block_id: int, dst_block_id: int,
+                            end_position_in_block: int) -> None:
+    """Copy the leading tokens in 1 block from host buffer to device layer.
+
+    Args:
+        src_buffer (torch.Tensor): The source buffer on host, shape is 
+            (2, len(block_ids), page_size, ...remaining dims...) 
+        dst_layer (torch.Tensor): The destination layer on device, shape is
+            (2, num_vllm_blocks, page_size, ...remaining dims...)
+        src_block_id (int): The source block id to copy.
+        dst_block_id (int): The destination block id to copy.
+        end_position_in_block (int): The end position in the block to copy.
+    """
+    dst_k = dst_layer[0][dst_block_id][:end_position_in_block]
+    src_k = src_buffer[0][src_block_id][:end_position_in_block]
+    dst_v = dst_layer[1][dst_block_id][:end_position_in_block]
+    src_v = src_buffer[1][src_block_id][:end_position_in_block]
+    dst_k.copy_(src_k, non_blocking=True)
+    dst_v.copy_(src_v, non_blocking=True)
+
+
+def h2d_copy_trailing_tokens(src_buffer: torch.Tensor, dst_layer: torch.Tensor,
+                             src_block_id: int, dst_block_id: int,
+                             start_position_in_block: int) -> None:
+    """Copy the trailing tokens in 1 block from host buffer to device layer.
+
+    Args:
+        src_buffer (torch.Tensor): The source buffer on host, shape is 
+            (2, len(block_ids), page_size, ...remaining dims...) 
+        dst_layer (torch.Tensor): The destination layer on device, shape is
+            (2, num_vllm_blocks, page_size, ...remaining dims...)
+        src_block_id (int): The source block id to copy.
+        dst_block_id (int): The destination block id to copy.
+        start_position_in_block (int): The start position in the block to copy.
+    """
+    dst_k = dst_layer[0][dst_block_id][start_position_in_block:]
+    src_k = src_buffer[0][src_block_id][start_position_in_block:]
+    dst_v = dst_layer[1][dst_block_id][start_position_in_block:]
+    src_v = src_buffer[1][src_block_id][start_position_in_block:]
+    dst_k.copy_(src_k, non_blocking=True)
+    dst_v.copy_(src_v, non_blocking=True)
```

Comment:
I think these functions are redundant?

I think can just change the type of `start_position_in_block` and `end_position_in_block` args in `h2d_copy_part_block` below from `int` to `Optional[int]` and use that for all cases?